From d4110568f5f526e4e2e35af027b657d0a7c5694c Mon Sep 17 00:00:00 2001
From: Eelco Dolstra <edolstra@gmail.com>
Date: Fri, 27 Jun 2025 17:59:15 +0200
Subject: [PATCH] Fix eval caching for path flakes

This fix is somewhat hacky since the accessor/input fingerprint
situation is a bit of a mess.
---
 src/libfetchers/fetchers.cc       | 6 ++++--
 src/libfetchers/path.cc           | 3 +--
 tests/functional/flakes/flakes.sh | 1 +
 3 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/libfetchers/fetchers.cc b/src/libfetchers/fetchers.cc
index 9cb896601..c42d83906 100644
--- a/src/libfetchers/fetchers.cc
+++ b/src/libfetchers/fetchers.cc
@@ -335,8 +335,10 @@ std::pair<ref<SourceAccessor>, Input> Input::getAccessorUnchecked(ref<Store> sto
 
     auto [accessor, result] = scheme->getAccessor(store, *this);
 
-    assert(!accessor->fingerprint);
-    accessor->fingerprint = result.getFingerprint(store);
+    if (!accessor->fingerprint)
+        accessor->fingerprint = result.getFingerprint(store);
+    else
+        result.cachedFingerprint = accessor->fingerprint;
 
     return {accessor, std::move(result)};
 }
diff --git a/src/libfetchers/path.cc b/src/libfetchers/path.cc
index 9239fd274..188ff27c0 100644
--- a/src/libfetchers/path.cc
+++ b/src/libfetchers/path.cc
@@ -128,8 +128,6 @@ struct PathInputScheme : InputScheme
 
         auto absPath = getAbsPath(input);
 
-        Activity act(*logger, lvlTalkative, actUnknown, fmt("copying %s to the store", absPath));
-
         // FIXME: check whether access to 'path' is allowed.
         auto storePath = store->maybeParseStorePath(absPath.string());
 
@@ -138,6 +136,7 @@ struct PathInputScheme : InputScheme
 
         time_t mtime = 0;
         if (!storePath || storePath->name() != "source" || !store->isValidPath(*storePath)) {
+            Activity act(*logger, lvlTalkative, actUnknown, fmt("copying %s to the store", absPath));
             // FIXME: try to substitute storePath.
             auto src = sinkToSource([&](Sink & sink) {
                 mtime = dumpPathAndGetMtime(absPath.string(), sink, defaultPathFilter);
diff --git a/tests/functional/flakes/flakes.sh b/tests/functional/flakes/flakes.sh
index ce695a6cb..2fe54b790 100755
--- a/tests/functional/flakes/flakes.sh
+++ b/tests/functional/flakes/flakes.sh
@@ -361,6 +361,7 @@ nix build -o $TEST_ROOT/result git+file://$flakeGitBare
 mkdir -p $flake5Dir
 writeDependentFlake $flake5Dir
 nix flake lock path://$flake5Dir
+[[ "$(nix flake metadata path://$flake5Dir --json | jq -r .fingerprint)" != null ]]
 
 # Test tarball flakes.
 tar cfz $TEST_ROOT/flake.tar.gz -C $TEST_ROOT flake5
-- 
2.49.0

